### ISSUE TITLE
keytab entries created with wrong salt

### ISSUE THREAD

**Issue by oflebbe** [msktutil/msktutil#66]
msktutil creates entries with wrong salt under some circumstances:

1) --use-service-account and existing account has no userPrincipalName. In that case the sAMACountname is taken for salt creation -- but in lowercase. Example: Administrator@EXAMPLE.COM has the salt EXAMPLE.COMAdministrator, but msktutil uses EXAMPLE.COMadministrator which leads to wrong keys

2) Enterprise names will probably not work for --use-service-account (not tested)

Reported by: mproehl

**Comment by mproehl** [comment]
Here is what happens:

The test account MaxM has no userPrincipalName attribute:
```
# ldapsearch -QLLL samaccountname=MaxM userPrincipalName sAMAccountName
dn: CN=Max Mustermann,OU=msktutil,DC=ads,DC=example,DC=com
sAMAccountName: MaxM

```

The correct salt uses the account name in mixed case (MaxM):
```
# KRB5_TRACE=/dev/stderr kinit MaxM@ADS.EXAMPLE.COM < /dev/null 2>&1 | grep salt
[20038] 1502786259.46475: Selected etype info: etype aes256-cts, salt "ADS.EXAMPLE.COMMaxM", params ""
# 
```

msktutil uses the lowercase account name (maxm):
```
# msktutil create --use-service-account --keytab /tmp/maxm.keytab --account-name MaxM --verbose 2>&1 | grep salt
 -- get_salt: Using salt of ADS.EXAMPLE.COMmaxm
# 
```

The resulting keys won't work except the arcfour key which don't use a salt

**Comment by michael-o** [comment]
Here it is, broken for enterprise principals, though the mechanism is properly described in `windows-salt.txt`:

My iUPN is `osipovmi@AD001.SIEMENS.NET` whereas my explicit UPN (enterprise principal) is `michael.osipov@siemens.com`. Look what happens with 6c768e3d6a6cbf90b6f6333a15809e269ba8262f:

     ./msktutil update --use-service-account --dont-change-password --keytab osipovmi.keytab --verbose --account-name osipovmi
     -- execute: Skipping creation of new password
     -- get_dc_host: Attempting to find Domain Controller to use via DNS SRV record in domain AD001.SIEMENS.NET for procotol tcp
     -- validate: Found DC: dc.ad001.siemens.net. Checking availability...
     -- get_dc_host: Found preferred Domain Controller: dc.ad001.siemens.net
     -- create_fake_krb5_conf: Created a fake krb5.conf file: /tmp/.msktkrb5.conf-OLokmK
     -- destroy_g_context: Destroying Kerberos Context
     -- initialize_g_context: Creating Kerberos Context
     -- finalize_exec: SAM Account Name is: osipovmi
     -- try_machine_password: Trying to authenticate for osipovmi with password
     -- create_default_machine_password: Default machine password for osipovmi is osipovmi
     -- try_machine_password: Error: krb5_get_init_creds_keytab failed (Preauthentication failed)
     -- try_machine_password: Authentication with password failed
     -- try_user_creds: Checking if default ticket cache has tickets
     -- finalize_exec: Authenticated using method 5
     -- LDAPConnection: Connecting to LDAP server: dc.ad001.siemens.net
    SASL/GSSAPI authentication started
    SASL username: osipovmi@AD001.SIEMENS.NET
    SASL SSF: 56
    SASL data security layer installed.
     -- ldap_get_base_dn: Determining default LDAP base: dc=AD001,dc=SIEMENS,dc=NET
     -- get_default_ou: Determining default OU: CN=Users,DC=ad001,DC=siemens,DC=net
     -- ldap_check_account: Checking that a service account for osipovmi exists
     -- ldap_check_account: Checking service account - found
     -- ldap_check_account: Found userAccountControl = 0x200
     -- ldap_check_account: Found supportedEncryptionTypes = 28
     -- ldap_check_account: Found User Principal: michael.osipov
     -- ldap_check_account_strings: Inspecting (and updating) service account attributes
     -- ldap_set_supportedEncryptionTypes: No need to change msDs-supportedEncryptionTypes they are 28
     -- ldap_set_userAccountControl_flag: Setting userAccountControl bit at 0x200000 to 0x0
     -- ldap_set_userAccountControl_flag: userAccountControl not changed 0x200
     -- ldap_get_kvno: KVNO is 5
     -- remove_keytab_entries: Trying to remove entries for osipovmi from keytab
     -- add_keytab_entries: Trying to add missing entries for osipovmi to keytab
     -- update_keytab: Updating all entries for osipovmi
     -- add_principal_keytab: Adding principal to keytab: osipovmi
     -- get_salt: Using salt of AD001.SIEMENS.NETmichael.osipov
     -- add_principal_keytab:   Adding entry of enctype 0x17
     -- add_principal_keytab:   Adding entry of enctype 0x11
     -- add_principal_keytab:   Adding entry of enctype 0x12
     -- add_principal_keytab: Adding principal to keytab: michael.osipov
     -- get_salt: Using salt of AD001.SIEMENS.NETmichael.osipov
     -- add_principal_keytab:   Adding entry of enctype 0x17
     -- add_principal_keytab:   Adding entry of enctype 0x11
     -- add_principal_keytab:   Adding entry of enctype 0x12

with written keytab:

    $ klist -k -t osipovmi.keytab     Keytab name: FILE:osipovmi.keytab
    KVNO Timestamp           Principal
    ---- ------------------- ------------------------------------------------------
       5 15.08.2017 10:40:59 osipovmi@AD001.SIEMENS.NET
       5 15.08.2017 10:40:59 osipovmi@AD001.SIEMENS.NET
       5 15.08.2017 10:40:59 osipovmi@AD001.SIEMENS.NET
       5 15.08.2017 10:40:59 michael.osipov@AD001.SIEMENS.NET
       5 15.08.2017 10:40:59 michael.osipov@AD001.SIEMENS.NET
       5 15.08.2017 10:40:59 michael.osipov@AD001.SIEMENS.NET

    $ kinit -k -t osipovmi.keytab michael.osipov@AD001.SIEMENS.NET
    kinit: Client 'michael.osipov@AD001.SIEMENS.NET' not found in Kerberos database while getting initial credentials
    
So, something is still fishy about it.

**Comment by mproehl** [comment]
Note: machine accounts use the lowercase account name (without trailing "$") as part of the salt:
``` 
# ldapsearch -QLLL samaccountname=LX02$ userPrincipalName sAMAccountName
dn: CN=LX02,CN=Computers,DC=ads,DC=example,DC=com
sAMAccountName: LX02$

# msktutil create --use-service-account --keytab /tmp/maxm.keytab --account-name LX02$ --verbose 2>&1 | grep salt
 -- get_salt: Using salt of ADS.EXAMPLE.COMlx02
# 
```

**Comment by mproehl** [comment]
the case sensitiveness issue is fixed with pull request #103 

the enterprise name issue is now #104 

closing this issue
